examples/torch/pearl_ml1_push.py

```diff
@@ -0,0 +1,163 @@
+"""PEARL ML1 example."""
+
+import akro
+from metaworld.benchmarks import ML1
+import numpy as np
+
+from garage.envs import normalize
+from garage.envs.base import GarageEnv
+from garage.envs.env_spec import EnvSpec
+from garage.experiment import LocalRunner, run_experiment
+from garage.experiment.meta_evaluator import MetaEvaluator
+from garage.experiment.task_sampler import SetTaskSampler
+from garage.sampler import LocalSampler
+from garage.torch.algos import PEARL
+from garage.torch.algos.pearl import PEARLWorker
+from garage.torch.embeddings import MLPEncoder
+from garage.torch.policies import ContextConditionedPolicy
+from garage.torch.policies import TanhGaussianMLPPolicy
+from garage.torch.q_functions import ContinuousMLPQFunction
+import garage.torch.utils as tu
+
+params = dict(
+    num_epochs=1000,
+    num_train_tasks=50,
+    num_test_tasks=10,
+    latent_size=7,
+    net_size=300,
+    meta_batch_size=16,
+    num_steps_per_epoch=4000,
+    num_initial_steps=4000,
+    num_tasks_sample=15,
+    num_steps_prior=750,
+    num_extra_rl_steps_posterior=750,
+    num_steps_per_eval=450,
+    batch_size=256,
+    embedding_batch_size=64,
+    embedding_mini_batch_size=64,
+    max_path_length=150,
+    reward_scale=10.,
+    use_information_bottleneck=True,
+    use_next_obs_in_context=False,
+    use_gpu=True,
+)
+
+
+def run_task(snapshot_config, *_):
+    """Set up environment and algorithm and run the task.
+
+    Args:
+        snapshot_config (garage.experiment.SnapshotConfig): The snapshot
+            configuration used by LocalRunner to create the snapshotter.
+            If None, it will create one with default settings.
+        _ : Unused parameters
+
+    """
+    # create multi-task environment and sample tasks
+    env_sampler = SetTaskSampler(lambda: GarageEnv(
+        normalize(ML1.get_train_tasks('push-v1'))))
+    env = env_sampler.sample(params['num_train_tasks'])
+
+    test_env_sampler = SetTaskSampler(lambda: GarageEnv(
+        normalize(ML1.get_test_tasks('push-v1'))))
+    #test_env = test_env_sampler.sample(params['num_train_tasks'])
+
+    runner = LocalRunner(snapshot_config)
+    obs_dim = int(np.prod(env[0]().observation_space.shape))
```

Comment:
is there a way we can hide this stuff inside the primitives or the algorithms? it seems it should be constant for every PEARL experiment.

Reply:
I need `obs_dim` and `action_dim` to make `env_spec` for qf and vf. I dont think there's a way around that right now.

Reply:
how about a classmethod on PEARL which takes the regular env_spec and returns the augmented one? that way, you can be guaranteed that you get the right spec.

Reply:
Are you suggesting to construct qf and vf inside PEARL as well? Because I need to pass the augmented space into qf and vf.

Reply:
I'm suggesting something like

```python
# g/t/a/pearl.py
class PEARL:

    @classmethod
    def augment_env_spec(cls, env_spec, latent_dim):
        # calculations...

        return EnvSpec(aug_obs, aug_act)

    @classmethod
    def get_encoder_spec(cls, env_spec, latent_dim):
        # calculations...

        return InOutSpec(in_space, out_space)
```

```python
# g/_dtypes.py
class InOutSpec:
    """Describes the input and output spaces of a primitive or module.

    Args:
        input_space (akro.Space): Input space of a module.
        output_space (akro.Space): Output space of a module.
    """
    def __init__(self, input_space, output_space):
        self._input_space = input_space
        self._output_space = output_space

    @property
    def input_space(self):
        """akro.Space: Input space of the module."""
        return self._input_space

    @property
    def output_space(self):
    """akro.Space: Output space of the module."""
        return self._output_space

    @property
    def transpose(self):
    """InOutSpec: A spec which has the input of this space as output, and vice-versa."""
        return InOutSpec(self._output_space, self._input_space)
```

```python
# g/e/env_spec.py
class EnvSpec(InOutSpec):
    """Describes the action and observation spaces of an environment.

    Args:
        observation_space (akro.Space): Observation space of the environment.
        action_space (akro.Space): Action space of the environment.
    """
    def __init__(self, observation_space, action_space):
        super().__init__(action_space, observation_space)

    @property
    def action_space(self):
    """akro.Space: Action space of the environment."""
        return self.input_space

    @property
    def observation_space(self):
    """akro.Space: Observation space of the environment."""
        return self.output_space
```